Remove usage of deprecated OSAtomic methods

LumioseSil opened the pull request:
#### 468a3a547bd29641a2a57ce21756e5f030884cda
<pre>
Remove usage of deprecated OSAtomic methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=251865">https://bugs.webkit.org/show_bug.cgi?id=251865</a>

Reviewed by Michael Catanzaro.

We should be using WTF::Atomics instead.

* Source/WebCore/platform/ios/wak/WebCoreThread.mm: Remove
  libkern/OSAtomics import.
* Source/WebKitLegacy/mac/WebView/WebView.mm: Make didDrawTiles a proper
  BOOL and replace OSAtomicCompareAndSwap32 with
  WTF::atomicCompareExchangeStrong.
* Source/WebKitLegacy/mac/WebView/WebViewData.h: Make didDrawTiles a
  proper BOOL.

Canonical link: <a href="https://commits.webkit.org/282479@main">https://commits.webkit.org/282479@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7faa8a2729d005f64147c4775ea0856cbd3cdcf0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63225 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42581 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15822 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67246 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13833 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65345 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50269 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14113 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50943 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9554 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66294 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39536 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54745 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31624 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36221 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12091 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12705 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57762 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12421 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68942 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7172 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12038 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58255 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7203 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54813 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58476 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14013 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5977 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38402 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39481 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40593 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39224 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/Atomics.h:
```diff
@@ -207,6 +207,12 @@ inline T atomicCompareExchangeStrong(T* location, T expected, T newValue, std::m
+inline T atomicCompareExchangeStrong(T* location, T expected, T newValue, std::memory_order order_success, std::memory_order order_failure)
+{
+    return bitwise_cast<Atomic<T>*>(location)->compareExchangeStrong(expected, newValue, order_success, order_failure);
+}
```
I don't understand this. Why do you need a new overload of atomicCompareExchangeStrong when you use the same memory order for both success and failure? Why not just use the version that takes only one memory order and already exists?

That said, it might not matter, because...

mcatanzaro commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -2267,10 +2267,8 @@ - (void)_dispatchTileDidDraw:(CALayer*)tile
 
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    if (!OSAtomicCompareAndSwap32(0, 1, &_private->didDrawTiles))
+    if (!WTF::atomicCompareExchangeStrong(&_private->didDrawTiles, NO, YES, std::memory_order_relaxed, std::memory_order_relaxed))
```
...although I'm not confident I understand memory order well enough, and although I also see you preserved the preexisting behavior here of not using any memory fence, I suspect the existing code was wrong and you actually _do_ require a memory barrier here, right? I think you need `std::memory_order_acq_rel` in the success case and `std::memory_order_acquire` in the failure case.

Having any mention of memory order makes it more difficult to reason about the correctness of the code, so it might be best to optimize for readability instead of performance and just use the default sequential consistent memory order, even if it's stricter than what is truly needed.

mcatanzaro commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -3431,7 +3429,7 @@ - (void)_setInitiatedDrag:(BOOL)initiatedDrag
 {
     if (frame == [self mainFrame])
-        _private->didDrawTiles = 0;
+        WTF::atomicStore(&_private->didDrawTiles, NO, std::memory_order_relaxed);
```
I think this should use `std::memory_order_release` (or the default sequential consistent order).

AZero13 replied:
Fixed!

mcatanzaro reviewed (changes requested):
Sorry for the delay. Using Atomics.h is a nice improvement. I'm not sure about using the relaxed memory order, though (even though this matches the existing code).

Also, I think the previous non-atomic store in _didCommitLoadForFrame invalidates the atomicity guarantees of other atomic operations on that variable. That's worth mentioning in your commit message, because you've fixed an actual bug, not just replaced the deprecated function call.

mcatanzaro reviewed (commented):
OK, I don't normally approve macOS/iOS pull requests, but surely this is fine.

I'm not sure how to retrigger EWS without updating the pull request, though. You might have to do one last force push, now that I've removed the skip-ews label.

LumioseSil committed:
Remove usage of deprecated OSAtomic methods
https://bugs.webkit.org/show_bug.cgi?id=251865

Reviewed by Michael Catanzaro.

We should be using WTF::Atomics instead.

* Source/WebCore/platform/ios/wak/WebCoreThread.mm: Remove
  libkern/OSAtomics import.
* Source/WebKitLegacy/mac/WebView/WebView.mm: Make didDrawTiles a proper
  BOOL and replace OSAtomicCompareAndSwap32 with
  WTF::atomicCompareExchangeStrong.
* Source/WebKitLegacy/mac/WebView/WebViewData.h: Make didDrawTiles a
  proper BOOL.

Canonical link: https://commits.webkit.org/282479@main
